### ISSUE TITLE
Strange behaviour of very trivial toy example

### ISSUE THREAD

**Issue by AlessandroFlati** [KindXiaoming/pykan#99]
Here's the gist showing a commented jupyter notebook demonstrating how KAN is failing at a very simple task, presumably because involving large numbers as outputs.
https://gist.github.com/AlessandroFlati/a5e6c5868bc4d67d51e99a7a5ec63322

**Comment by AlessandroFlati** [comment]
> large numbers as 'inputs'?
> 
> you need to do necessary 'data preprocess'(better) or 'normizer layer'.
> 
> you can rember scaling-factor and later scale-back.

Thanks for the explanation, but I'm definitely not new to Deep Learning and ML in general. First of all I called them outputs, not inputs,, since inputs here should be considered the numbers in the range you're choosing in the dataset creation, but if you read carefully enough, the question should be clear.

Here we're facing an exponential of a rational function, with a large bias. The bias is leant almost immediately, and it doesn't (or shouldn't, at least) affect the choice of the variable part. Backpropagation for sure helps in finding that bias (that's why the `[1,1]` example works), but each layer it's backpropagated without its bias in the spline form. Here we know the real solution can be expressed in machine-precision as a `[1,1,1,1]` - a composition of `x`, `1/x`, `exp` - but I'm guessing it's either the rational function itself or the large biases of each function that makes the backpropagation go `nan`. The last part, though, doesn't explain the odd behavior explained in the gist.

@KindXiaoming Any more educated idea from you?

**Comment by KindXiaoming** [comment]
Hi yeah it's always good to apply normalization. Please correct me if I'm wrong, but I think neural networks (include both MLPs and KANs) should always be dealing with normalized data. Please try MLPs as well, I guess MLPs will get similar poor performance on your unnormalized data.

**Comment by yuedajiong** [comment]
I don't have any ability to offer more educated ideas, but I've taken a look at your function and made a plot. 
Within my understanding, I think I've figured out the reason: including normalization but not limited.


![f_n5](https://github.com/KindXiaoming/pykan/assets/52232153/14a33778-2984-4217-895f-55ff1790f94d)

![f_n1000](https://github.com/KindXiaoming/pykan/assets/52232153/cf4ae225-4a49-4cf2-b367-11490e63c23c)

[fuck_large.zip](https://github.com/KindXiaoming/pykan/files/15228508/fuck_large.zip)

**Comment by AlessandroFlati** [comment]
@yuedajiong I think you're missing the point (apart from presenting data in a strange way).

![image](https://github.com/KindXiaoming/pykan/assets/12647648/7b4c45e9-610d-4d4f-813b-7e8bf703999b)

@KindXiaoming Yes, everything regarding backpropagation in a neural network generally suffers from large numbers. I'm not even complaining about this: I deal with these aspects everyday.
The fact is that as a mathematician and software engineer, I strongly believe that KANs will reveal themselves not prune to these normalization error, since the spline convolution happens on a separate "level", and without singularities should be perfectly able to reach SOTA results on par with symbolic regression offered from Gauss-Newton method, Levenberg-Marquardt algorithm, etc.
For now, I'll close this since it's probably dealing with too many things at once, but I still think this can be a good test in the future.

**Comment by AlessandroFlati** [comment]
Just to clarify from a software perspective: the expectation here is that KAN is able to match (or leverage the use of) scipy `curve_fit`, when the auto symbolic fixes all layers.

**Comment by yuedajiong** [comment]
I won't engage in this issue anymore, and I've also deleted my reply because I found my posted is a very friendly suggestion, while yours always seem unfriendly. 

A final piece of friendly advice: Apart from the large number/exponential issues you mentioned, try setting train_num=10, then print these values sequentially to see. When using plt.plot(x, y), just do it directly without additional parameters, don't use dots for plotting, as you only need the values and their total number. Try using lines instead, it will be easier for you to observe the sequence.

"presenting data in a strange way", maybe not 'strange', but 'better'?

You constructed a messy and complex function with a default value of 1000, which, when viewed as continualization, is essentially a single-input multiple-output function. Even KAN or modern NNs cannot fit it; even if KAN could fit such a functional form, it couldn't be fitted with just two or three layers with a very small grid. 

Whether my understanding is correct or not, there's no need to reply.


[fuck_large.py.txt](https://github.com/KindXiaoming/pykan/files/15234760/fuck_large.py.txt)

**Comment by yuedajiong** [comment]
I'm not entirely sure if I fully understand what you mean.  
I can only roughly estimate what you're trying to express from your limited text and the functions in your code.

By plotting the function(I didn't use dots, so what you see along the lower right edge is basically dots) and looking at the parameters in your function, I found that the exp(...) part actually ranges approximately from 0 to 1.6. Then you multiply it by a large number 3669.4, and add an even larger number 7.588e7, and see if KAN can fit well/correctlly. 
I'm not very rigorous. I'll try to convert inputs and outputs to have a mean of 0 and a range of [-1, +1]，maybe wider, and meaningful representations won't be very small floating-point numbers.
I think these are some considerations in the engineering implementation. It might be feasible to subtract the mean and then divide by the range; you can simultaneously observe x and y to ensure that the input-output falls within a "stable" range compatible with KAN.   

If you are specifically testing the numerical stability of KAN's implementation, then there's not much else to say.

Regarding numerical stability, besides being very large, you can also construct a value that is small enough, which might not fit well either.
Including precision variances due to float-number representations in different intervals, the sufficient number of terms in expansions for certain functions like sine(.), etc.
Expanding numerical ranges and ensuring numerical stability are very fine engineering problems. Even relatively mature products like PyTorch and even GLIBC are continuously improving in this regard.

If you're talking about the approximation capability and precision at algorithmic levels, such as MLPs, KAN/Extendable-B-Spline, or other methods like Taylor expansion, Fourier series, then you, being a mathematician, have more authority to speak on the matter.

That's all what I've been trying to say from the first response until now.

**Comment by AlessandroFlati** [comment]
Thanks for the kind reply.
Yes, I was more interested in the fact that there are many other methods to reach an `R^2 = 1` approximation of this function, which have little to no problem handling ver large/little numbers. Of course I understand the engineering limit of backpropagation, which gradients can explode or vanish even for smaller numbers than the proposed example ones - but I wouldn't expect this to happen in a KAN.

Since there already are implementations of such approximation methods, maybe we could strive to incorporate them in the learning loop, so that even before pruning and finetuning the suggestions made by the symbolic engine are more accurate or even at level machine, in a negligible computational time. In order to do so, though, we would have to plan a "reversing" method that given a symbolic formula generates an appropriate KAN that can be later finetuned. 
This could very well, imho, drastically reduce the time difference that actually exists between MLPs and KANs.
